### PR TITLE
Emit a build operation describing task inputs and build cache key

### PR DESCRIPTION
**Pull request by facewindu**
This change isolates the TaskExecuters that are part of the computation of the task artifact state and the task output caching state, in order to run them in a BuildOperation.
The result of this build operation is effectively the `TaskOutputCachingBuildCacheKey`, and is consumed by the build scan plugin.

This has been spec'd in https://github.com/gradle/dotcom/blob/master/docs/specs/capturing-task-input-properties-and-build-cache-key.md

See CI 'Stage 2 - Linux Basic Coverage' on my branch: https://builds.gradle.org/viewLog.html?buildId=2903626

### CODE REVIEW

**Comment by facewindu** [comment]
@alkemist do you want to weigh in ?
My implementation complexifies `TaskExecutionServices.createTaskExecuter()`, but the was to be as in the spec, and get just what was needed in the build operation.

**Comment by wolfs** [comment]
@facewindu What I don't understand in the spec why we want to capture all the loading of the task artifact state, too. This happens always, not just when the build cache is enabled and could be wrapped in an own build operation.
What I was expecting was wrapping the actual calculation into a build operation. This would mean wrapping a BuildOperation around [these two lines](https://github.com/gradle/gradle/blob/ff1dc7fc80eac707bf8222ae4b2a960631933d58/subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveBuildCacheKeyExecuter.java#L42-L43).
What would be the downside of this?

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveBuildCacheKeyExecuter.java [hunk 1]
```diff
@@ -21,33 +21,79 @@
...
+            an explicit step that always happens earlier and wrap it.
+            Regardless, it would be good to formalise the input work in some form so it doesn't just
+            happen as a side effect of calling some method for the first time.
+         */
+        buildOperationExecutor.run(new RunnableBuildOperation() {
+            @Override
+            public void run(BuildOperationContext buildOperationContext) {
+                TaskOutputCachingBuildCacheKey cacheKey = doResolve(task, context);
+                buildOperationContext.setResult(cacheKey);
+                context.setBuildCacheKey(cacheKey);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveBuildCacheKeyExecuter.java]: Let's clarify the name of this `context` a bit, given that there's another `context` in this, well, context...

Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveBuildCacheKeyExecuter.java]: I'll do this on `release` instead. Let's merge this.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveTaskArtifactStateTaskExecuter.java [hunk 2]
```diff
@@ -48,12 +48,6 @@ public void execute(TaskInternal task, TaskStateInternal state, TaskExecutionCon
         context.setTaskArtifactState(taskArtifactState);
         outputs.setHistory(taskArtifactState.getExecutionHistory());
         LOGGER.info("Putting task artifact state for {} into context took {}.", task, clock.getElapsed());
-        try {
-            executer.execute(task, state, context);
-        } finally {
-            outputs.setHistory(null);
-            context.setTaskArtifactState(null);
-            LOGGER.debug("Removed task artifact state for {} from context.");
-        }
```
Comment (facewindu) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveTaskArtifactStateTaskExecuter.java]: this is now moot given we went back to the old decorator pattern, with a subsection of the task executers being wrapped in a build operation

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveTaskOutputCachingStateExecuter.java [hunk 3]
```diff
@@ -51,6 +52,8 @@ public void execute(TaskInternal task, TaskStateInternal state, TaskExecutionCon
         } else {
             state.setTaskOutputCaching(DefaultTaskOutputs.DISABLED);
         }
-        delegate.execute(task, state, context);
+        if (delegate != null) {
+            delegate.execute(task, state, context);
+        }
```
Comment (facewindu) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveTaskOutputCachingStateExecuter.java]: this is now moot given we went back to the old decorator pattern, with a subsection of the task executers being wrapped in a build operation

**Pull request merged**